CONTRIBUTING.md

```diff
@@ -1,18 +1,149 @@
+   ```
+
+1. Install BinderHub:
+
+   ```bash
+   python3 -m pip install -e .
+   ```
+
+1. Install the JavaScript dependencies:
+
+   ```bash
+   npm install
+   ```
+
+1. Create the JS and CSS bundles with:
+
+   ```bash
+   npm run webpack
+   ```
+  Note: you need to run this every time you make a change to the CSS or JS
+  for it to take effect.
+
+1. Run it!
+
+   ```bash
+   python3 -m binderhub -f testing/localonly/binderhub_config.py
+   ```
+
+1. Visit http://localhost:8585 to see it in action.
+
+Building and launching repositories will not work. You can still work on the
+user interface of those parts as BinderHub is configured to fake those
+actions. You can tell you are using the fake builder and launcher from the fact
+that the build will never complete.
+
+To learn how to set yourself with a BinderHub development environment that
+lets you modify the builder and launcher refer to
+[Kubernetes integration changes](#Kubernetes-integration-changes).
+
+
+## Kubernetes integration changes
+
+Setting yourself up to make changes to the kubernetes integration of BinderHub
+requires a few one-time setup steps. These steps are described in the
+"One-time installation" section below. Follow those first then return here for
+day to day development procedures.
+
+
+### Day to day development tasks
+
+After having setup minikube and helm once, these are the tasks you need for
+every day development.
+
+* Start and stop minikube with `minikube start` and `minikube stop`.
+* Install JupyterHub in minikube with helm `./testing/minikube/install-hub`
+* Setup docker to use the same docker daemon as your minikube cluster `eval $(minikube docker-env)`
+* Start BinderHub `python3 -m binderhub -f testing/minikube/binderhub_config.py`
+* Visit your BinderHub at[http://localhost:8585](http://localhost:8585)
+
+To execute most of our test suite you need a running minikube cluster.
+It does not need to have anything installed on it though:
+
+```bash
+minikube start
+pytest -svx -m "not auth_test"
+```
+
+The tests should generate familiar pytest like output and complete in a few
+seconds.
+
+To execute all the main tests use `./ci/test-main` which will setup a
+JupyterHub on minikube for you. These tests will generate a lot of output and
+take a few minutes to run. The tests will attempt to clean up after themselves
+on your minikube cluster.
+
+To execute the tests related to authentication use `./ci/test-auth` which will
+setup a JupyterHub on minikube for you and use configuration files to configure
+your BinderHub to require authentication. These tests will generate a lot of
+output and take a few minutes to run. The tests will attempt to clean up after
+themselves on your minikube cluster.
+
+
+### One-time installation
+
+To run the full BinderHub it needs to have access to a kubernetes cluster
+with a JupyterHub installed on it. This is what we will setup in this section.
 
 Before you begin, there are a few utilities that need to be installed:
```

Comment:
May be worth mentioning that `apt` is Linux based and provide alternatives for Mac users?

Reply:
Let me know what you think of the sentence I added above. When we originally wrote the guide we discussed about using conda or system python, OSX commands etc and the conclusion then was that we would give the Ubuntu commands that are as simple as possible and assume that if you dislike them or want to deviate because you have opinions you'd be able to do so.

For example virtualenvs/conda envs don't feature but are a good idea IMHO. However writing instructions for that opinionated stack (and the not opinionated version) would be a lot of work.

What do you think of this approach and an we make a nice one sentence version of all of the above to explain this to people?

Reply:
Looks great! I think the clarification that `apt` is Ubuntu based is enough. In my naivety, I was very confused that I couldn't execute the first steps until it was explained to me that the command wasn't _supposed_ to work for me :)

Reply:
Merci.